js/data/bucket/symbol_bucket.js

```diff
@@ -333,6 +340,46 @@ SymbolBucket.prototype.addFeature = function(lines, shapedText, shapedIcon, feat
     }
 };
 
+SymbolBucket.prototype.findPolygonAnchors = function(line, lines, polygons) {
+
+    if (line.length === 0) {
+        return [];
+    } else if (line.length < 3) {
+        return [ new Anchor(line[0].x, line[0].y, 0) ];
+    }
+
+    var anchors = null;
+    // Work out what line is...
+    var polygon = null;
+    var isPolygon = false;
+    var isHole = false;
+    for (var i = 0; !isPolygon && !isHole && i < polygons.length; i++) {
+        polygon = polygons[i];
+        isPolygon = line === polygon[0];
+        if (!isPolygon) {
+            for (var j = 1; !isHole && j < polygon.length; j++) {
+                isHole = line === polygon[i];
+            }
+        }
+    }
```

Comment:
It looks like for each polygon ring, you go through all other rings doing equality comparisons, which is O(n^2). However it does not seem to be necessary — you could instead add symbol instances as you loop through the rings above. This would need some code reorganization above.